src/main/scala/firrtl/transforms/FlattenRegUpdate.scala

```diff
@@ -58,38 +58,61 @@ object FlattenRegUpdate {
     * @return [[firrtl.ir.Module Module]] with register updates flattened
     */
   def flattenReg(mod: Module): Module = {
-    // We want to flatten Mux trees for reg updates into if-trees for
-    // improved QoR for conditional updates.  However, unbounded recursion
-    // would take exponential time, so don't redundantly flatten the same
-    // Mux more than a bounded number of times, preserving linear runtime.
-    // The threshold is empirical but ample.
-    val flattenThreshold = 4
-    val numTimesFlattened = mutable.HashMap[Mux, Int]()
-    def canFlatten(m: Mux): Boolean = {
-      val n = numTimesFlattened.getOrElse(m, 0)
-      numTimesFlattened(m) = n + 1
-      n < flattenThreshold
-    }
+    // We want to flatten Mux trees for reg updates into if-trees for improved
+    // QoR for conditional updates. However, when-otherwise structure often
```

Comment:
It is not clear to me what you mean by `when-otherwise structure`. Maybe you could rephrase this comment to require less background information on how register updates are emitted.

Reply:
I'll try to rephrase, it's not about how their emitted, it's about what last-connect semantics mean, given:
```scala
when (io.a) {
  reg  := foo
  when (!io.b)
    reg := bar
  }
}
// Semantically, at this point the "fallback" value for reg is:
//   mux(io.a, mux(!io.b, bar, foo), reg)
when (io.b) {
  when (io.c) {
    when (io.d) {
      reg := fizz
    }
  }
}
```
The problem being solved is that the 2nd when structure has 3 "holes" where the complex fallback currently (before this PR) gets inlined. The improvement in this PR is detecting this case and not inlining it.

Reply:
Updated with an example, let me know what you think